Step3 - 구간 추가 기능

mins99 opened the pull request:
안녕하세요 리뷰어님!
step2의 피드백을 반영하고 step3 구현하여 리뷰 요청 드립니다.
step2가 빠르게 머지되어서 ATDD 미션 금방 끝나겠구나 했는데.. step3 구현이 정말 정말 너무 어려웠습니다 😭

3단계 미션을 진행한 내용을 정리하면 다음과 같습니다 😊

1. IllegalArgumentException 예외 처리를 추가하였습니다
2. Section List를 가지는 Sections 를 추가하였습니다
   1) 구간 추가에 대한 유효성 검증은 Sections 에서 합니다
   2) 구간 추가시 기존 구간 사이인지에 대한 검증 및 수정은 Section 에서 합니다.
3. distance 원시값을 포장하였습니다
4. 구간 추가에 대한 인수 테스트를 예외 케이스를 포함하여 작성했습니다

꼼꼼하게 구현하려고는 노력했는데 그래도 부족함이 많습니다~ 이번 단계도 잘 부탁드립니다!! 🙏🏻

mins99 committed:
refactor(RequestDTO): RequestEntity 이름 변경

mins99 committed:
refactor(ResponseDTO): 구간 정보를 가져오는 로직 수정

- LineResponse 에서 Response 객체를 가지도록 변경
- 변경이 일어나지 않도록 각 객체의 변수에 final 선언

mins99 committed:
test(LineAcceptanceTest): 지하철 노선 생성 테스트 로직 수정

- LineRequest에 추가했던 테스트를 위한 코드를 삭제하면서 테스트에 임시 로직을 추가함

mins99 committed:
fix(Line): 지하철 노선 생성시 response 데이터 오류 수정

- 구간 정보를 중복으로 넣고 있던 부분 수정

mins99 committed:
docs(README): 3단계 - 구간 추가 기능 요구 사항 정리

mins99 committed:
test(AcceptanceTest): BeforeEach 메서드 공통 클래스에서 호출하도록 수정

mins99 committed:
refactor(DatabaseCleanup): 패키지 분리

mins99 committed:
feat(controller): IllegalArgumentException 예외 처리 추가

mins99 committed:
refactor(Section): 변수명 변경 및 원시값 포장

- stationid -> station
- distance 객체 생성

mins99 committed:
chore(AcceptanceTest): 패키지 경로 변경

mins99 committed:
feat(SectionRequest): 지하철 구간 요청 객체 추가

mins99 committed:
fix(Response): 생성자 접근 제어자 public -> private 변경

mins99 committed:
feat(Distance): 거리 객체 유효성 체크 로직 수정 및 기능 추가

- 거리 조회 메서드 추가
- 구간 추가시 거리 값 조정 메서드 추가

mins99 committed:
feat(Section): 상행역, 하행역 정보 수정 메서드 추가

mins99 committed:
feat(Sections): 구간(Section) 리스트를 가지는 일급 컬렉션 객체 추가

- 구간 추가시 유효성 검증
- 역 사이에 새로운 역을 등록하는 경우 구간 정보 및 거리 수정

mins99 committed:
refactor(Line): Section list를 일급 컬렉션으로 변경

- 역 조회 로직도 개선해봄

mins99 committed:
feat(saveSection): 구간 추가 기능 구현

mins99 committed:
test(RestAssuredTemplate): path parameter를 가지는 post 메서드 추가

mins99 committed:
test(SectionAcceptanceTest): 지하철 구간 인수 테스트 작성

mins99 committed:
chore(Line): 미사용 패키지 삭제

enemfk777 commented on src/main/java/nextstep/subway/domain/Distance.java:
```diff
@@ -0,0 +1,33 @@
+        }
+    }
+
+    public void changeDistance(long distance) {
+        this.distance -= distance;
+    }
```
Distance를 인자로 받고,
새로운 Distance 인스턴스를 반환해서 객체의 불변성을 지켜보는건 어떨까요?

mins99 replied:
오 새로운 인스턴스를 반환하면 거리 변경시 값이 0이하인 예외에 대해서도 처리가 가능해지겠네요. 수정해보겠습니다~

enemfk777 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,69 @@
+                        && x.getLineStations().contains(section.getDownStation()));
+    }
+
+    private boolean isLongerThanStation(Section section) {
+        return sections.stream()
+                .filter(x -> x.getUpStation().equals(section.getUpStation())
+                        || x.getDownStation().equals(section.getDownStation()))
+                .anyMatch(x -> x.getDistance() <= section.getDistance());
+    }
```
저장된 section과 추가될 section 간의 거리 validation은 
sections의 책임일까요 section의 책임일까요?

mins99 replied:
`findTargetSection`의 로직과 중복되는 부분이 있어서 고민했었는데 책임의 대상을 Section으로 변경하니 해당 로직에 대해 삭제가 가능하겠네요! section으로 변경하겠습니다 😊

enemfk777 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,69 @@
+
+    private boolean isContainAnyStation(Section section) {
+        return sections.stream()
+                .anyMatch(x -> x.getLineStations().contains(section.getUpStation())
```
x와 같이 너무 축약한 이름, section과 같이 너무 구분이 안가는 이름 보다는
currentSection, newSection 과 같이 비즈니스로직상의 의미를 잘 나타내는 이름은 어떨까요?

mins99 replied:
넵 의미를 나타내는 이름으로 수정하겠습니다!

enemfk777 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,69 @@
+        return sections;
+    }
+
+    public void addSection(Section section) {
+        if (!sections.isEmpty()) {
+            validSectionCheck(section);
+            updateSection(section);
+        }
+        sections.add(section);
+    }
```
Section, Sections 등 중요한 도메인 객체들이 추가되었는데,
도메인 클래스의 유닛 테스트도 작성해보시면 어떨까요? :)

mins99 replied:
도메인 테스트 로직은 신경을 안썼네요 T_T 도메인 테스트도 추가해보겠습니다~

enemfk777 commented on src/test/java/nextstep/subway/station/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,160 @@
+                new SectionRequest(upStationId, downStationId, distance));
+    }
+
+    void 지하철_노선에_지하철_구간_등록됨(ExtractableResponse<Response> response) {
```
등록 검증이 조금 빈약한것 같아요 :)
등록 후 조회 하면 등록 된 역이 올바른 순서로 (e.g. 강남 - 잠실 사이에 삼성을 추가하면 강남 - 삼성 -잠실 순 반환)
조회되는지까지 함께 검증해보면 어떨까요?

mins99 replied:
등록후에 조회하면 순서가 전혀 안맞았는데 해당 부분도 수정하고 테스트 추가하겠습니다!

enemfk777 reviewed (changes requested):
안녕하세요 민성님 🙇‍♂️

3단계 열심히 구현해주셨네요 👍
이번 단계는 어려운게 정상입니다 😁
저도 예전에 이 미션에 시간을 많이 쏟았던 기억이 나네요 :)

잘 구현 해주셨지만, 몇 가지 아쉬운 부분이 있어 코멘트 드렸으니
반영 후 다시 요청 부탁드립니다!

mins99 committed:
refactor(Sections): 변수명 변경 및 리스트 조회 로직 수정

- 너무 축약하거나 의미를 알 수 없는 변수명을 쓰기 보다는 비즈니스 로직상 의미를 잘 나타내는 이름을 사용하자
- 실제 구간 순서대로 조회하도록 수정

mins99 committed:
refactor(Distance): 비교 로직 및 검증 로직 위치 변경

- 거리 validation은 Section에서 distance에게 메시지를 던진다
- changeDistance의 매개변수를 Distance로 받고, 새로운 Distance 인스턴스를 반환해서 객체의 불변성을 지키자

mins99 committed:
refactor(All): 패키지 구조 변경

mins99 committed:
refactor(All): 패키지 구조 변경

mins99 committed:
test(SectionAcceptanceTest): 테스트 구체화

- 구간 등록 시 의도대로 구간이 등록되었는지 검증 추가

mins99 committed:
feat(domain): 도메인 유닛테스트를 위한 생성자 추가

mins99 committed:
refactor(Section): 거리 변경시 유효성 검증 추가

mins99 committed:
test(All): 도메인 및 레포지토리 테스트 추가

mins99 commented:
안녕하세요 리뷰어님!
피드백 주신 내용 바탕으로 수정 진행해보았습니다~

수정한 사항은 다음과 같습니다
1. 거리에 대한 유효성 검증 로직을 Sections -> Section 으로 변경하였습니다
2. 스트림에서 사용한 변수명에 대해 의미있는 이름으로 변경하였습니다
3. 노선 구간 조회시 실제 구간 순서대로 조회 되도록 수정하였습니다
4. 등록에 대한 인수테스트를 조금 더 명확하도록 개선하였습니다
5. 도메인 및 레포지토리에 대한 테스트를 추가하였습니다

ATDD 미션이라는 생각으로 도메인 테스트는 신경을 쓰지 않았는데,
도메인 테스트를 추가하면서 로직 오류도 발견했네요 ㅎㅎ; 인수테스트 뿐만 아니라 단위 테스트의 중요성을 다시 느꼈습니다!

추가로, 단위 테스트도 추가하게 되면서 테스트 패키지에 대한 고민이 생겼는데요..ㅠㅠ
리뷰어님은 ATDD와 TDD 등등 다양한 케이스(?)의 테스트를 추가하시면 어떤 기준으로 패키지를 구분하시는지 궁금합니다!

확인 부탁드립니다~ 🥰

enemfk777 commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -11,7 +12,9 @@ public class Sections {
     public List<Section> getSections() {
-        return sections;
+        return sections.stream()
+                .sorted((s1, s2) -> s1.getUpStation().getName().compareTo(s2.getDownStation().getName()))
```
정렬 조건이 어색한 것 같아요 :)
String 의 compareTo 는 사전순 정렬이 될 텐데
지금 출력해야하는 순서가 사전순 정렬은 아닐 것 같아요 :)

enemfk777 commented on src/test/java/nextstep/subway/domain/DistanceTest.java:
```diff
@@ -0,0 +1,59 @@
+        assertThat(distance.getDistance()).isEqualTo(1);
+    }
+
+    @ParameterizedTest(name = "Distance 객체는 0미만의 수를 인자로 받으면 오류를 반환한다")
```
테스트의 명칭은 구체적인 구현을 나타내기보단
요구사항을 나타내는 방향으로 작성해보시는건 어떨까요?
(e.g. Distance 객체는 0미만의 수를 인자로 받으면 오류를 반환한다 -> 구간 거리는 1 이상이어야 한다.)
요구사항 정리서가 있을때,
테스트코드명칭이 요구사항과 같이 적혀있다면 어떤 장점이 있을까요? :)

mins99 replied:
로또 미션 진행시에는 각 도메인 별 요구사항을 정리했었는데, 요구사항을 상세히 정리하지 않다보니 구현에 대한 내용을 쓰게 된것 같습니다
어떤 요구사항의 검증을 위해 만든 테스트인지 명시하는 것이 다른 개발자나, 다른 직군들과 협업할 때 좋을 것 같습니다. 
해당 부분도 신경써서 작성해보겠습니다!

enemfk777 commented on src/test/java/nextstep/subway/domain/LineTest.java:
```diff
@@ -0,0 +1,45 @@
+    @Test
+    @DisplayName("getLineStations의 결과 상행종착역 부터 하행종착역 순서대로 반환된다")
+    void getLineStations() {
+        // given
+        Line line = new Line("2호선", "green");
+        line.addSection(section1);
+        line.addSection(section2);
+
+        // when
+        List<Station> actual = line.getLineStations();
+
+        // then
+        assertThat(actual).containsExactly(station1, station2, station3);
```
```
// given
Line line = new Line("2호선", "green");
Station 삼성역 = new Station(1L, "삼성역");
Station 잠실역 = new Station(2L, "잠실역");
Station 강변역 = new Station(3L, "강변역");;
Section 잠실_강변 = new Section(1L, 잠실역, 강변역, 2);
Section 삼성_잠실 = new Section(2L, 삼성역, 잠실역, 5);
line.addSection(잠실_강변);
line.addSection(삼성_잠실);

// when
List<Station> actual = line.getLineStations();

// then
assertThat(actual).containsExactly(삼성역, 잠실역, 강변역);
```
다른 코멘트에서 말씀드린 어색한 정렬방법으로 인해 위와 같은 케이스는 통과하지 못하고 있는 것 같아요 :)

mins99 replied:
제가 운좋게 가나다순으로 등록을 했었네요...😅 수정하겠습니다

enemfk777 reviewed (approved):
안녕하세요 민성님:)
빠르게 피드백 반영해주셧네요 !
도메인 테스트코드 까지 추가해주시느라 고생 많으셨습니다.
다만, 정렬 방식이 요구사항과 다른 것 같아 코멘트 남겨드렸으니,
이 부분은 다음단계에서 함께 반영 부탁드립니다!!

이번단계는 이만 머지하겠습니다 :) 고생하셨습니다 🙇‍♂️
